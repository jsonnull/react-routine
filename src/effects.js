// @flow

export type Effect =
  | { type: 'SET_STATE', state: Object }
  | { type: 'FORCE_UPDATE' }
  | { type: 'RENDER' }
  | { type: 'COMPONENT_WILL_MOUNT' }
  | { type: 'COMPONENT_DID_MOUNT' }
  | { type: 'COMPONENT_WILL_UNMOUNT' }
  | { type: 'SHOULD_COMPONENT_UPDATE', callback: () => boolean }
  | { type: 'COMPONENT_WILL_UPDATE' }
  | { type: 'COMPONENT_DID_UPDATE' }
  | { type: 'COMPONENT_DID_CATCH' }
  | { type: 'CALLBACK', name: string }
  | { type: 'CREATE_CALLBACKS', callbacks: { [name: string]: Function } }

export const SET_STATE = 'SET_STATE'
export const FORCE_UPDATE = 'FORCE_UPDATE'
export const RENDER = 'RENDER'
export const COMPONENT_WILL_MOUNT = 'COMPONENT_WILL_MOUNT'
export const COMPONENT_DID_MOUNT = 'COMPONENT_DID_MOUNT'
export const COMPONENT_WILL_UNMOUNT = 'COMPONENT_WILL_UNMOUNT'
export const SHOULD_COMPONENT_UPDATE = 'SHOULD_COMPONENT_UPDATE'
export const COMPONENT_WILL_UPDATE = 'COMPONENT_WILL_UPDATE'
export const COMPONENT_DID_UPDATE = 'COMPONENT_DID_UPDATE'
export const COMPONENT_DID_CATCH = 'COMPONENT_DID_CATCH'
export const CALLBACK = 'CALLBACK'
export const CREATE_CALLBACKS = 'CREATE_CALLBACKS'

export const setState = (state: Object) => ({
  type: SET_STATE,
  state: state
})

export const forceUpdate = () => ({
  type: FORCE_UPDATE
})

export const render = () => ({
  type: RENDER
})

export const componentWillMount = () => ({
  type: COMPONENT_WILL_MOUNT
})

export const componentDidMount = () => ({
  type: COMPONENT_DID_MOUNT
})

export const componentWillUnmount = () => ({
  type: COMPONENT_WILL_UNMOUNT
})

export const shouldComponentUpdate = (callback: () => boolean) => ({
  type: SHOULD_COMPONENT_UPDATE,
  callback
})

export const componentWillUpdate = () => ({
  type: COMPONENT_WILL_UPDATE
})

export const componentDidUpdate = () => ({
  type: COMPONENT_DID_UPDATE
})

export const componentDidCatch = () => ({
  type: COMPONENT_DID_CATCH
})

export const oneOf = (...args: Array<Effect>) => args

export const createHandlers = (callbacks: { [name: string]: Function }) => ({
  type: CREATE_CALLBACKS,
  callbacks
})

export const handler = (name: string) => ({
  type: CALLBACK,
  name
})
