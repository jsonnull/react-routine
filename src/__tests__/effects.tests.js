// @flow
import * as effects from '../effects'

describe('effects', () => {
  it('should create a forceUpdate effect', () => {
    expect(effects.forceUpdate()).toMatchSnapshot()
  })

  it('should create a render effect', () => {
    expect(effects.render()).toMatchSnapshot()
  })

  it('should create a componentWillMount effect', () => {
    expect(effects.componentWillMount()).toMatchSnapshot()
  })

  it('should create a componentDidMount effect', () => {
    expect(effects.componentDidMount()).toMatchSnapshot()
  })

  it('should create a componentWillUnmount effect', () => {
    expect(effects.componentWillUnmount()).toMatchSnapshot()
  })

  it('should create a componentWillUpdate effect', () => {
    expect(effects.componentWillUpdate()).toMatchSnapshot()
  })

  it('should create a componentDidUpdate effect', () => {
    expect(effects.componentDidUpdate()).toMatchSnapshot()
  })

  it('should create a componentDidCatch effect', () => {
    expect(effects.componentDidCatch()).toMatchSnapshot()
  })

  /*
   * Some more complex effects
   */

  it('should create a setState effect', () => {
    const mockState = { test: 'testing' }
    expect(effects.setState(mockState)).toEqual({
      type: 'SET_STATE',
      state: mockState
    })
  })

  it('should create a shouldComponentUpdate effect', () => {
    const mockCallback = () => true
    expect(effects.shouldComponentUpdate(mockCallback)).toEqual({
      type: 'SHOULD_COMPONENT_UPDATE',
      callback: mockCallback
    })
  })

  it('should create a oneOf effect', () => {
    expect(
      effects.oneOf(
        effects.componentWillMount(),
        effects.componentWillUnmount()
      )
    ).toMatchSnapshot()
  })

  it('should create a createHandler effect', () => {
    const mockHandlers = { handlerName: () => 'test' }
    expect(effects.createHandlers(mockHandlers)).toEqual({
      type: 'CREATE_CALLBACKS',
      callbacks: mockHandlers
    })
  })

  it('should create a handler effect', () => {
    const mockHandlers = { handlerName: () => 'test' }
    expect(effects.handler('handlerName')).toEqual({
      type: 'CALLBACK',
      name: 'handlerName'
    })
  })
})
