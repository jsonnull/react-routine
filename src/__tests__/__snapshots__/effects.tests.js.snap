// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effects should create a componentDidCatch effect 1`] = `
Object {
  "type": "COMPONENT_DID_CATCH",
}
`;

exports[`effects should create a componentDidMount effect 1`] = `
Object {
  "type": "COMPONENT_DID_MOUNT",
}
`;

exports[`effects should create a componentDidUpdate effect 1`] = `
Object {
  "type": "COMPONENT_DID_UPDATE",
}
`;

exports[`effects should create a componentWillMount effect 1`] = `
Object {
  "type": "COMPONENT_WILL_MOUNT",
}
`;

exports[`effects should create a componentWillUnmount effect 1`] = `
Object {
  "type": "COMPONENT_WILL_UNMOUNT",
}
`;

exports[`effects should create a componentWillUpdate effect 1`] = `
Object {
  "type": "COMPONENT_WILL_UPDATE",
}
`;

exports[`effects should create a forceUpdate effect 1`] = `
Object {
  "type": "FORCE_UPDATE",
}
`;

exports[`effects should create a oneOf effect 1`] = `
Array [
  Object {
    "type": "COMPONENT_WILL_MOUNT",
  },
  Object {
    "type": "COMPONENT_WILL_UNMOUNT",
  },
]
`;

exports[`effects should create a render effect 1`] = `
Object {
  "type": "RENDER",
}
`;
